lib/lightning/kafka_triggers/pipeline.ex

```diff
@@ -0,0 +1,212 @@
+defmodule Lightning.KafkaTriggers.Pipeline do
+  @moduledoc """
+  Broadway pipeline that processes messages from Kafka clusters and persists
+  the received messages if they are not duplicating a previous message.
+  """
+  use Broadway
+
+  alias Ecto.Multi
+  alias Lightning.KafkaTriggers
+  alias Lightning.KafkaTriggers.TriggerKafkaMessage
+  alias Lightning.KafkaTriggers.TriggerKafkaMessageRecord
+  alias Lightning.Repo
+  alias Lightning.Workflows.Trigger
+
+  require Logger
+
+  def start_link(opts) do
+    trigger_id = opts |> Keyword.get(:trigger_id)
+
+    Broadway.start_link(__MODULE__,
+      name: trigger_id,
+      context: %{
+        trigger_id: trigger_id
+      },
+      producer: [
+        module: {
+          BroadwayKafka.Producer,
+          build_producer_opts(opts)
+        },
+        concurrency: 1
+      ],
+      processors: [
+        default: [
+          concurrency: 10
+        ]
+      ],
+      batchers: []
+    )
+  end
+
+  @impl true
+  def handle_message(_processor, message, context) do
+    %{trigger_id: trigger_id} = context
+
+    %{
+      data: data,
+      metadata: %{
+        key: key,
+        offset: offset,
+        partition: partition,
+        topic: topic,
+        ts: timestamp
+      }
+    } = message
+
+    topic_partition_offset =
+      KafkaTriggers.build_topic_partition_offset(message)
+
+    record_changeset =
+      TriggerKafkaMessageRecord.changeset(
+        %TriggerKafkaMessageRecord{},
+        %{
+          topic_partition_offset: topic_partition_offset,
+          trigger_id: trigger_id |> Atom.to_string()
+        }
+      )
+
+    message_changeset =
+      %TriggerKafkaMessage{}
+      |> TriggerKafkaMessage.changeset(%{
+        data: data,
+        key: key,
+        message_timestamp: timestamp,
+        metadata: message.metadata,
+        offset: offset,
+        topic: topic,
+        trigger_id: trigger_id |> Atom.to_string()
+      })
+
+    trigger_changeset =
+      Trigger
+      |> Repo.get(trigger_id |> Atom.to_string())
+      |> Trigger.kafka_partitions_changeset(partition, timestamp)
+
+    Multi.new()
+    |> Multi.insert(:record, record_changeset)
+    |> Multi.insert(:message, message_changeset)
+    |> Multi.update(:trigger, trigger_changeset)
```

Comment:
I would prefer that the trigger object would be a result only from the UI, from user input, but I understand the straightforward approach to centralize the config in one place.

Reply:
@jyeshe Noted - the goal of this pass was to keep the footprint as small as possible, but there are already plans to  separate the cluster config from the consumer config and we can certainly extend that to have the tracking of partition timestamps be a new table as well. That way, we can isolate the operational updates.